native_client/deepspeech.cc

```diff
@@ -103,28 +124,66 @@ Model::infer(float* aMfcc, int aNFrames, int aFrameLen)
   std::vector<Tensor> outputs;
   Status status = mPriv->session->Run(
     {{ "input_node", input }, { "input_lengths", n_frames }},
-    {"output_node"}, {}, &outputs);
+    {"Reshape_3"}, {}, &outputs);
   if (!status.ok()) {
     std::cerr << "Error running session: " << status.ToString() << "\n";
     return NULL;
   }
 
+  auto outputs_mapped = outputs[0].tensor<float, 3>();
+  // The CTCDecoder works with log-probs.
+  for (int t = 0; t < timesteps; ++t) {
+    for (int b = 0; b < batch_size; ++b) {
+      for (int c = 0; c < num_classes; ++c) {
+        input_data_mat[t][b][c] = outputs_mapped(t, b, c);
+      }
+    }
+  }
+
+  CTCBeamSearchDecoder<KenLMBeamState> decoder(num_classes,
+                              /* beam width */ mPriv->beam_width,
+                                               mPriv->scorer,
+                              /* batch size */ 1);
+
+  // Raw data containers (arrays of floats, ints, etc.).
+  int sequence_lengths[batch_size] = {timesteps};
+
+  // Convert data containers to the format accepted by the decoder, simply
+  // mapping the memory from the container to an Eigen::ArrayXi,::MatrixXf,
+  // using Eigen::Map.
+  Eigen::Map<const Eigen::ArrayXi> seq_len(&sequence_lengths[0], batch_size);
+  std::vector<Eigen::Map<const Eigen::MatrixXf>> inputs;
+  inputs.reserve(timesteps);
+  for (int t = 0; t < timesteps; ++t) {
+    inputs.emplace_back(&input_data_mat[t][0][0], batch_size, num_classes);
+  }
+
+  // Prepare containers for output and scores.
+  // CTCDecoder::Output is std::vector<std::vector<int>>
+  std::vector<CTCDecoder::Output> decoder_outputs(top_paths);
+  for (CTCDecoder::Output& output : decoder_outputs) {
+    output.resize(batch_size);
+  }
+  float score[batch_size][top_paths] = {{0.0}};
+  Eigen::Map<Eigen::MatrixXf> scores(&score[0][0], batch_size, top_paths);
+
+  decoder.Decode(seq_len, inputs, &decoder_outputs, &scores).ok();
```

Comment:
I'm reviewing my code, that I do find wonderful :)

Reply:
Technically you're reviewing TensorFlow test code :)